Add de-duplicating apps, dbs, and desktops when sorting/totalCount is needed

kimlisa opened the pull request:
fixes https://github.com/gravitational/teleport/issues/12555

#### Description
Retrieving resources from the web api layer requires resources to be de-duplicated per page fetch to get accurate `totalCount` and accurately fill out `liimit` request

- PR adds de-duplicating matches while filtering for `FakePaginate` func. This is the path that the web UI will always take. This handles de-duplicating for `apps`, `databases`, `desktops`, and `kubes`. 
  - This change does makes `FakePaginate` less efficient b/c it will always run filter for every resource per fetch versus just running it on the initial fetch to get the total count. But this method was required to handle sorting + dedup feature and handle cases where resource names were the same, but labels were different. 
- PR removes inaccurate de-dupping in the web api layer since we do it from the back now
- PR also adds another criteria when de-duplicating `applications` by checking for its `public_addr` as well as its name. So apps with same name but different public addresses will not be considered a duplicate

kimlisa committed:
Add deduplication for app,dbs,desktops in fake paginating

kimlisa committed:
Remove deduplicating in api web layer as its done in the back

kimlisa committed:
Remove old test, add sort test in auth layer

kimlisa committed:
Revert presence file

kimlisa committed:
Add deduplication in presence layer
 - removed unused SortBy field in windows desktop request

kimlisa commented on lib/services/local/presence.go:
```diff
@@ -1559,7 +1559,7 @@ func (s *PresenceService) DeleteAllWindowsDesktopServices(ctx context.Context) e
 	switch {
 	case req.RequiresFakePagination():
-		return s.listResourcesWithSort(ctx, req)
+		return nil, trace.BadParameter("request requiring getXXX resource methods are not supported in presence.ListResources")
```
i didn't think this path will ever be used (bypassing auth with roles layer), so i thought it would be nice to have one less thing to maintain, but i can see how it can cause confusion. i reverted it back and also placed de-duplication and tests

kimlisa commented:
friendly ping @rosstimothy @espadolini :pray:

espadolini commented on lib/auth/auth_with_roles.go:
```diff
@@ -1211,6 +1211,7 @@ func (a *ServerWithRoles) listResourcesWithSort(ctx context.Context, req proto.L
 		if err := servers.SortByCustom(req.SortBy); err != nil {
 			return nil, trace.Wrap(err)
 		}
+		servers = types.DeduplicateAppServers(servers)
```
For this (and the other ones): why sort and then dedup, rather than dedup then sort?

espadolini reviewed (approved):
I wonder if it's worth it to add a special case for sort-by-name-and-dedup as a single step (because then the dedup can be done with a single pass, just by skipping consecutive duplicates). Maybe if it turns out to be a bottleneck?

rosstimothy reviewed (commented):
Are nodes affected by #12555 too? If so why aren't they deduplicated too?

`(*ServerWithRoles) listResourcesWithSort` and `(*PresenceService) listResourcesWithSort` look nearly identical. Is there any reason when a request requires fake pagination that  `(*ServerWithRoles) listResourcesWithSort` is called directly instead of `(*PresenceService) ListResources` which would then call `(*PresenceService) listResourcesWithSort`?

kimlisa commented on api/types/appserver.go:
```diff
@@ -377,3 +377,18 @@ func (s AppServers) GetFieldVals(field string) ([]string, error) {
 }
+
+// DeduplicateAppServers deduplicates appservers with the same application name.
+func DeduplicateAppServers(servers []AppServer) []AppServer {
```
i don't think so, b/c the way we are dedupping on the proxy side, we don't account for expiry, do you think it matters?

kimlisa committed:
Address CR

kimlisa commented:
> > > Are nodes affected by #12555 too? If so why aren't they deduplicated too?
> > 
> > 
> > no, node names are unique
> 
> You could potentially have duplicate nodes in the event that an old node isn't TTLed out fast enough(looking at you dynamodb).

i'm most likely missing something, but i didn't understand how nodes could be dupped because we use a `put` operation in the backend: https://github.com/gravitational/teleport/blob/5ec15cbbb9ddce7b7746de37e0cd6a8ed5fa78fd/lib/services/local/presence.go#L323

so there seems to be only ever a 1-1 relationship where one node is one ssh service?

espadolini commented:
The current behavior in the web ui will not deduplicate ssh names, but you can easily have duplicates if you just run two nodes with the same `nodename` (happens when clearing the datadir and reregistering, for instance).

kimlisa committed:
Address edge case concerns

- Resources can have same name but can have different labels:
  FakePaginatie will now deduplicate while applying filters.
- Applications can have same name, but can also have different public
  addresses: when deduplicating apps, use both name and address as unique key

kimlisa commented:
@espadolini @rosstimothy I addressed the edge case concerns that came up on discussion on slack last week: https://github.com/gravitational/teleport/pull/12685/commits/616ef396b1fed3014c8ee3bf54acde4d30c81eca

In order for the de-duplication to work correctly with the sorting feature, the `FakePaginate` func will be less efficient in that every fetch will filter every resources now. In the web UI, this was already being applied on the first fetch's anyways (to get the total count)
  - this handles filtering for resources with the same name but with different labels, so we apply the filter first, then discard the rest
  - when de-dupping applications, we now account for different public addresses with the name

tested manually with shells

kimlisa commented on api/types/app.go:
```diff
@@ -301,14 +301,16 @@ func (a *AppV3) CheckAndSetDefaults() error {
 	seen := make(map[string]struct{})
 	for _, app := range apps {
-		if _, ok := seen[app.GetName()]; ok {
+		key := fmt.Sprintf("%s%s", app.GetName(), app.GetPublicAddr())
```
these are great edge case questions i never thought about, and i'm not sure it was ever accounted for. the scope for this requires how dedupping is currently working for the CLI as well (i think), and i spoke with @zmb3 on my 1x1 and he said this can be done in another issue where while we dedup we check `if dup, check for later expiry date`, is this okay?

espadolini commented on api/types/app.go:
```diff
@@ -301,14 +301,16 @@ func (a *AppV3) CheckAndSetDefaults() error {
-// DeduplicateApps deduplicates apps by name.
+// DeduplicateApps deduplicates apps by combination of app name and public address.
+// Apps can have the same name but also could have different addresses.
 func DeduplicateApps(apps []Application) (result []Application) {
```
Are we ok with `tsh` now getting duplicated app names as a result of this? And what about the other `DeduplicateFoo` functions?

kimlisa committed:
Address CR: add benchmark and use custom struct for apps

kimlisa commented on api/types/app.go:
```diff
@@ -301,14 +301,16 @@ func (a *AppV3) CheckAndSetDefaults() error {
+// DeduplicateApps deduplicates apps by combination of app name and public address.
+// Apps can have the same name but also could have different addresses.
 func DeduplicateApps(apps []Application) (result []Application) {
 	seen := make(map[string]struct{})
```
the others didn't apply since they only dedup on `name`

kimlisa commented on lib/cache/cache_test.go:
```diff
@@ -793,6 +793,62 @@ func benchListNodes(b *testing.B, nodeCount int, pageSize int) {
+cpu: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
+BenchmarkListResourcesWithSort-8               1        2351035036 ns/op
+*/
+func BenchmarkListResourcesWithSort(b *testing.B) {
```
hopefully i did this right, it was a replica of `benchGetNodes`: https://github.com/gravitational/teleport/blob/master/lib/cache/cache_test.go#L713

i ran this 10 times to get the average time

espadolini commented on lib/services/local/presence.go:
```diff
@@ -1622,40 +1622,45 @@ func FakePaginate(resources []types.ResourceWithLabels, req proto.ListResourcesR
-	for currIndex, resource := range resources {
-		switch match, err := services.MatchResourceByFilters(resource, filter); {
+	// Iterate and filter every resource, deduplicating while matching.
+	seenResourceMap := make(map[string]struct{})
```
Can we have this also be a hashset of pairs of strings? The additional string can just be empty.

zmb3 commented on lib/cache/cache_test.go:
```diff
@@ -742,6 +742,63 @@ func benchGetNodes(b *testing.B, nodeCount int) {
+
+	ctx := context.Background()
+
+	count := 100000
+	for i := 0; i < count; i++ {
```
```suggestion
	for i := 0; i < b.N; i++ {
```

zmb3 commented on lib/cache/cache_test.go:
```diff
@@ -742,6 +742,63 @@ func benchGetNodes(b *testing.B, nodeCount int) {
+
+	count := 100000
+	for i := 0; i < count; i++ {
+		func() {
```
Why do you need the anonymous function here as opposed to just running this code in the loop directly?

kimlisa replied:
tbh i didn't pay attention to it, i asked the original author and they don't remember, so i removed it

kimlisa committed:
Address CR: fix benchmark and use custom struct as key

kimlisa committed:
Address CR: add different range limit to benchmark

kimlisa committed:
Add clarifying comment

kimlisa committed:
Merge branch 'master' into lisa/bug/fix-total-count

kimlisa committed:
Merge branch 'master' into lisa/bug/fix-total-count

kimlisa committed:
Merge branch 'master' into lisa/bug/fix-total-count

kimlisa committed:
Merge branch 'master' into lisa/bug/fix-total-count

kimlisa committed:
Merge branch 'master' into lisa/bug/fix-total-count

zmb3 reviewed (approved):
Bot.

kimlisa committed:
Merge branch 'master' into lisa/bug/fix-total-count

kimlisa committed:
Merge branch 'master' into lisa/bug/fix-total-count
